### PR TITLE
[Shop][Checkout] Upgrade addressing form to a component

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A
|-----------------|-----
| Branch?         | bootstrap-shop
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutAddressingContext.php [hunk 1]
```diff
@@ -223,7 +223,7 @@ public function iSpecifiedTheBillingAddress(?AddressInterface $address = null)
      * @When I specify the email as :email
      * @When I do not specify the email
      */
-    public function iSpecifyTheEmail($email = null)
+    public function iSpecifyTheEmail($email = null): void
```
Comment (mpysiak) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutAddressingContext.php]:
```suggestion
    public function iSpecifyTheEmail(?string $email = null): void
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 2]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
...
+                configuration:
+                    type: 'billing'
+                    field: 'differentBillingAddress'
+            billing_address:
+                template: '@SyliusShop/checkout/address/form/addresses/address.html.twig'
+                configuration:
+                    type: 'billing'
+                    isDifferentAddress: true
+            
+
```
Comment (mpysiak) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]:
```suggestion
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 3]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
...
-                template: '@SyliusShop/checkout/address/form/addresses/address_book.html.twig'
+                component: 'sylius_shop:checkout:address:address_book'
+                props:
+                    field: 'billingAddress'
+            form:
+                template: '@SyliusShop/checkout/address/form/addresses/address/form.html.twig'
+                configuration:
+                    field: 'billingAddress'
+            
+
```
Comment (mpysiak) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]:
```suggestion
```

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php [hunk 4]
```diff
@@ -155,12 +154,13 @@ public function specifyBillingAddressFullName(string $fullName): void
 
     public function canSignIn(): bool
     {
-        return $this->waitForElement(5, 'login_button');
+        $this->waitForElementUpdate('form');
+        return $this->hasElement('login_button');
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php]:
```suggestion
        $this->waitForElementUpdate('form');

        return $this->hasElement('login_button');
```

File: src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php [hunk 5]
```diff
@@ -204,19 +208,20 @@ public function backToStore(): void
...
-        $this->waitForElement(5, 'shipping_province');
+        $this->waitForElementUpdate('form');
         $this->getElement('shipping_province')->setValue($provinceName);
     }
 
     public function hasShippingAddressInput(): bool
     {
-        return $this->waitForElement(5, 'shipping_province');
+        $this->waitForElementUpdate('form');
+        return $this->hasElement('shipping_province');
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php]:
```suggestion
        $this->waitForElementUpdate('form');

        return $this->hasElement('shipping_province');
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 6]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
                 template: '@SyliusShop/checkout/address/form/addresses.html.twig'
             navigation:
                 template: '@SyliusShop/checkout/address/form/navigation.html.twig'
-                
-        'sylius_shop.checkout.address.form.addresses':
+
+        'sylius_shop.checkout.address.form.addresses#billingRequired':
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]:
```suggestion
        'sylius_shop.checkout.address.form.addresses#billing_required':
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 7]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
...
+                configuration:
+                    type: 'shipping'
+                    field: 'differentShippingAddress'
+            shipping_address:
+                template: '@SyliusShop/checkout/address/form/addresses/address.html.twig'
+                configuration:
+                    type: 'shipping'
+                    isDifferentAddress: true
+        
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]:
```suggestion

```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 8]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
...
+        'sylius_shop.checkout.address.form.addresses#shippingRequired':
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]:
```suggestion
        'sylius_shop.checkout.address.form.addresses#shipping_required':
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml [hunk 9]
```diff
@@ -15,11 +18,66 @@ twig_hooks:
...
+                    type: 'billing'
+                    field: 'differentBillingAddress'
+            billing_address:
+                template: '@SyliusShop/checkout/address/form/addresses/address.html.twig'
+                configuration:
+                    type: 'billing'
+                    isDifferentAddress: true
+            
+        
+        'sylius_shop.checkout.address.form.addresses.address#billing':
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/checkout/address.yaml]: Shouldn't it be
```suggestion
        'sylius_shop.checkout.address.form.addresses.billing_address':
```
to be consistent with above hooks? 🤔

File: src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/AddressBookComponent.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+#[AsLiveComponent]
+class AddressBookComponent
+{
+    use DefaultActionTrait;
+    use ComponentToolsTrait;
+    use HookableLiveComponentTrait;
+
+    #[LiveProp(writable: true, onUpdated: 'addressUpdated')]
+    public ?int $address = null;
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/AddressBookComponent.php]: Int? Shouldn't it be:
```suggestion
    public ?int $addressId = null;
```
or even
```suggestion
    public mixed $addressId = null;
```
🤔

File: src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    #[PreReRender(priority: -100)]
+    public function checkEmailExist(): void
+    {
+        $email = $this->formValues['customer']['email'] ?? null;
+        if (null !== $email) {
+            if ($this->shopUserRepository->findOneByEmail($email)) {
+                $this->emailExists = true;
+            } else {
+                $this->emailExists = false;
+            }
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php]:
```suggestion
            $this->emailExists = $this->shopUserRepository->findOneByEmail($email) ? true : false;
```
it should probably work the same way 🤔

File: src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+
+    #[LiveListener('sylius:shop:address-updated')]
+    public function addressFieldUpdated(#[LiveArg] int $addressId, #[LiveArg] string $field): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php]:
```suggestion
    public function addressFieldUpdated(#[LiveArg] mixed $addressId, #[LiveArg] string $field): void
```

File: src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+        $this->formValues[$field] = $newAddress;
+    }
+
+    protected function instantiateForm(): FormInterface
+    {
+        return $this->formFactory->create(
+            $this->formClass,
+            $this->order,
+            ['customer' => $this->customerContext->getCustomer()]
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Twig/Component/Checkout/Address/FormComponent.php]:
```suggestion
            ['customer' => $this->customerContext->getCustomer()],
```

File: src/Sylius/Bundle/ShopBundle/templates/checkout/address/form/addresses/address/form.html.twig [hunk 14]
```diff
@@ -0,0 +1,6 @@
+{% set prefixes = hookable_metadata.prefixes|merge(['sylius_shop.shared.form']) %}
+{% set form = hookable_metadata.context.form %}
+{% set field = hookable_metadata.configuration.field %}
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/templates/checkout/address/form/addresses/address/form.html.twig]:
```suggestion

```

**Review by GSadee — APPROVED** [review]

**Review by TheMilek — APPROVED** [review]

**Pull request merged**